### ISSUE TITLE
Leaking/Too many open File descriptors in Konnectivity server with GRPC mode and UDS transport

### ISSUE THREAD

**Issue by rata** [kubernetes-sigs/apiserver-network-proxy#276]
Hi!

This investigation is joint work we are doing with @ipochi, I’m just opening the issue because it is late in his timezone now :)

We are trying to adopt Konnectivity v0.0.24 and we faced some memory issues when using it in GRPC mode with UDS (unix sockets). In production the mem usage of the konnectivity container was too high and restarted due to that, on several clusters. For now, we switched to http-connect mode and we are having more luck with that so far.

Nevertheless, we are trying to investigate the GRPC issue in a lab setup. For this, @ipochi created [a CLI tool][konnscen] using client-go to reproduce different scenarios. This just gets pod logs on some random pods, with different levels of concurrency.

Using that tool, we can easily reproduce a scenario where we see a lot of open fds on the konnectivity server, even when the proxy is idle for ~1 day. And if we hit it with more load, the number of fds seems to always increase (>= than before). We also see a big number of goroutines too. On a quick look it seems several might be per connection, to receive and such. But we haven’t dug deeper into that yet.

I’m using an example with a small number of fds here on what I will paste, but with more load we can easily open 2k+ fds and the same observations here are true when more fds are open too. One konnscen config.yaml that can be used for that is:
```
concurrent_connections:
  number_of_concurrent_users: 50
  number_of_times: 2
```

Taking a closer look into the fds, we can see that the fds are sockets connections. In this case, 60020 is the PID of the konnectivity proxy server:
```
# ls -l /proc/60020/fd
lrwx------. 1 root root 64 Oct 26 21:59 0 -> /dev/null
l-wx------. 1 root root 64 Oct 26 21:59 1 -> 'pipe:[203909]'
lrwx------. 1 root root 64 Oct 25 16:24 10 -> 'socket:[201974]'
lrwx------. 1 root root 64 Oct 25 16:24 11 -> 'socket:[204824]'
lrwx------. 1 root root 64 Oct 27 13:18 12 -> 'socket:[8739504]'
lrwx------. 1 root root 64 Oct 25 16:24 13 -> 'socket:[199329]'
lrwx------. 1 root root 64 Oct 27 13:18 16 -> 'socket:[8735695]'
lrwx------. 1 root root 64 Oct 27 13:18 19 -> 'socket:[8737576]'
l-wx------. 1 root root 64 Oct 26 21:59 2 -> 'pipe:[203910]'
lrwx------. 1 root root 64 Oct 27 13:18 20 -> 'socket:[8738266]'
lrwx------. 1 root root 64 Oct 27 13:18 21 -> 'socket:[8738267]'
lrwx------. 1 root root 64 Oct 27 13:18 24 -> 'socket:[8737581]'
lrwx------. 1 root root 64 Oct 27 13:18 25 -> 'socket:[8737584]'
lrwx------. 1 root root 64 Oct 27 13:18 29 -> 'socket:[8740168]'
lrwx------. 1 root root 64 Oct 25 16:23 3 -> 'anon_inode:[eventpoll]'
lrwx------. 1 root root 64 Oct 27 13:18 31 -> 'socket:[8739460]'
lrwx------. 1 root root 64 Oct 27 13:18 33 -> 'socket:[8739464]'
lrwx------. 1 root root 64 Oct 27 13:18 37 -> 'socket:[8740172]'
lr-x------. 1 root root 64 Oct 25 16:23 4 -> 'pipe:[204821]'
lrwx------. 1 root root 64 Oct 27 13:18 46 -> 'socket:[8740178]'
lrwx------. 1 root root 64 Oct 27 13:18 47 -> 'socket:[8740179]'
l-wx------. 1 root root 64 Oct 25 16:23 5 -> 'pipe:[204821]'
lrwx------. 1 root root 64 Oct 27 13:18 51 -> 'socket:[8736418]'
lrwx------. 1 root root 64 Oct 27 13:18 53 -> 'socket:[8738186]'
lrwx------. 1 root root 64 Oct 27 13:18 55 -> 'socket:[8739471]'
lrwx------. 1 root root 64 Oct 27 13:18 59 -> 'socket:[8739477]'
lrwx------. 1 root root 64 Oct 25 16:24 6 -> /var/konnectivity-server/konnectivity-server.log
lrwx------. 1 root root 64 Oct 25 16:24 7 -> 'socket:[198534]'
lrwx------. 1 root root 64 Oct 25 16:24 8 -> 'socket:[198541]'
lrwx------. 1 root root 64 Oct 25 16:24 9 -> 'socket:[204822]'
```
The lower fds are just stdin, out, err, the listen on the unix socket, some go runtime fds to use poll, the established connections with the agent, etc. Those are about the first 14-16 fds. After those, there are a lot of socket fds. When we increase the number of leaked fds, the number of socket fds is increased, the rest is more or less the same.

The number between square brackets is the inode allocated for that fd. When we search for this, we can see it is a connection accepted on the unix socket (this is also in the host, as the pod uses hostNetwork):

```
# grep 8738266 /proc/net/tcp* /proc/net/udp* /proc/net/unix
/proc/net/unix:ffff8ac1d1e26800: 00000003 00000000 00000000 0001 03 8738266 /var/konnectivity-server/konnectivity-server.socket
```

Using a script we verified that this is the case for almost all socket fds.
And looking with ss it indeed seems these sockets are connections with the APIserver on this same node:
```
# ss -a --unix -p | grep 8738266
u_str ESTAB  0      0                                                                                       * 8735729            * 8738266 users:(("kube-apiserver",pid=27932,fd=131))                                                                    
u_str ESTAB  0      0                                     /var/konnectivity-server/konnectivity-server.socket 8738266            * 8735729 users:(("proxy-server",pid=60020,fd=20))           
```

Looking at the code, it seems we do the listen on this socket [here][uds-listen] and then is just handled by [grpc here][grpc-serve]. So, indeed it seems the API server is establishing these connections and for some reason they are not closed.

The first thing we noticed is that runUDSFrontendServer() is [not setting any keepalive][no-keepalive] configuration on the gRPC server. In contrast, for tcp transport this is done in runMTLSFrontendServer [here][keepalive]. Setting that _seems_ to help, although it doesn’t completely solve the issue. But we need to run more tests to verify (it is quite easy to have misleading tests in this kind of issue). Also, the [keepalive.ServerParameters.Time param][grpc-keepalive-param] seems to send a PING command and that, docs I found, says it uses underneath the http/2 ping command. I’m not sure if http/2 is being used when we use UDS instead of tcp, so not sure if that param in the keepalive is a no-op when using UDS. We will verify, probably try to sniff the traffic and see if we see the ping and a connection closed due to that. If that is the case, we will open a PR with that :)

We also tried using the [keepalive.ServerParameters.MaxConnectionIdle][grpc-keepalive-param], but no strong conclusions just yet. We will report back and open a PR when we are certain that it improves the situation.

We also tried to use GRPC with tcp transport, as we thought that if this is a grpc issue, it is more unlikely to happen with a tcp transport that is so widely used in the wild. But when trying that, we hit a wall: it seems the [APIserver doesn’t support GRPC with tcp transport][apiserver-no-tcp].

Also, for these tests we upgraded the grpc version used. Currently [1.27 is used][grpc-version], that is more than 1.5 years old. We upgraded to 1.41.0 (it seems some leaks were fixed), latest, just in case. But it didn't help.

We also tried to see what these connections are doing, to gain some more understanding. And with this simple patch, we see that the connections from the APIserver seem open and healthy from konnectivity server POV. It _might_ be okay for grpc to not close them:

```
diff --git a/pkg/server/server.go b/pkg/server/server.go
index f8f7825..0f148ff 100644
--- a/pkg/server/server.go
+++ b/pkg/server/server.go
@@ -378,8 +378,10 @@ func (s *ProxyServer) Proxy(stream client.ProxyService_ProxyServer) error {
        }()
 
        // Start goroutine to receive packets from frontend and push to recvCh
+       // TODO rata: why create a goroutine for this? Why not just do it here?
        go func() {
                for {
+                       klog.V(2).ErrorS(fmt.Errorf("About to receive"), "userAgent", userAgent, "serverID", s.serverID)
                        in, err := stream.Recv()
                        if err == io.EOF {
                                klog.V(2).InfoS("Stream closed on Proxy", "userAgent", userAgent, "serverID", s.serverID)
@@ -399,7 +401,20 @@ func (s *ProxyServer) Proxy(stream client.ProxyService_ProxyServer) error {
                }
        }()
 
-       return <-stopCh
+       var ret error
+       t := time.Tick(1 * time.Minute)
+
+       for {
+               select {
+               case ret = <-stopCh:
+                       klog.V(2).ErrorS(fmt.Errorf("stoppping"), "XXX: rata. closing conn", "err", ret)
+                       break
+               case <-t:
+                       klog.V(2).ErrorS(fmt.Errorf("conn opened"), "XXX: rata. clock is ticking")
+               }
+       }
+
+       return ret
 }

```

With this patch, we see a lot of open grpc connections. In fact, we see a lot more than open fds. And at this time we realized that grpc multiplexes the connections and sends several over one fd (I think around 50 connections per fd is the default). In fact, the prometheus metric `konnectivity_network_proxy_server_grpc_connections{service_method="Proxy"}` shows the same value of connections we see if we grep the logs for the open connections (like running `cat /var/konnectivity-server/konnectivity-server.log | grep XXX | grep '16:39' | wc -l` we can get the open connections as we only filter for the ones printed in some minute. And, luckily, that matches :-D).

The setup we are using in our lab tests is: Konnectivity 0.0.24 on single master 3 worker node cluster on Equinix Metal using [Lokomotive](https://github.com/kinvolk/lokomotive). Setup instructions with manifests and all can be found [here][konn-loko], in case you want to look into some config detail.

So, I think this also leads us to investigate what is happening from the apiserver POV. I’m curious to see if the connections to the proxy-server are open there and if it detected the end user connection as broken or not, and if we can easily detect that.

Another thing to try is see if we can reproduce the behaviour with kubectl logs, instead of this custom client-go programs. We might be not closing the connection properly or something in konnscen, maybe.

Also, while the observed behavior of this and what we face in prod when using konnectivity seem to match, we have not confirmed yet it is the same. We will try to rollout again grpc with lot of care and some selected subset of users, to verify if the open fds and other metrics seem to be high and have more confidence that this issue is the same we faced in production. Will keep you posted :)

Another issue, that might be the same or very related (not sure yet), is that if we restart the proxy-server when there are already some “fd leaks”, it starts again with more open fds than before. All, again, connections from the API server. And a quick test seems to indicate that, if we wait some time to start connectivity, api server opens even more fds. It seems it retries while this is down and creates new connections. When proxy-server is up, all connections are established and we have more open fds. But this is something we need to investigate too… :)

As you can see, I think we have a good understanding of the issue, but still have several open questions :). Does any of this rings any bells for you? Do you have any pointers or tips to further debug this? Any ideas/suggestions are super welcome!

Also @cheftako this is the issue I commented during the SIG meeting last week :)

[konnscen]: https://github.com/ipochi/konnscen
[uds-listen]:https://github.com/kubernetes-sigs/apiserver-network-proxy/blob/4e8b12d2cd77db7ae5acecb5b706d5ecb3bf93e3/cmd/server/app/server.go#L150 
[grpc-serve]: https://github.com/kubernetes-sigs/apiserver-network-proxy/blob/4e8b12d2cd77db7ae5acecb5b706d5ecb3bf93e3/cmd/server/app/server.go#L179 
[no-keepalive]: https://github.com/kubernetes-sigs/apiserver-network-proxy/blob/4e8b12d2cd77db7ae5acecb5b706d5ecb3bf93e3/cmd/server/app/server.go#L173-L179 
[keepalive]: https://github.com/kubernetes-sigs/apiserver-network-proxy/blob/4e8b12d2cd77db7ae5acecb5b706d5ecb3bf93e3/cmd/server/app/server.go#L254 
[grpc-keepalive-param]: https://pkg.go.dev/google.golang.org/grpc@v1.41.0/keepalive#ServerParameters 
[apiserver-no-tcp]: https://github.com/kubernetes/kubernetes/blob/a48a2efbd45ad77901dd09f2665d8cc1e1d8dbf6/staging/src/k8s.io/apiserver/pkg/server/egressselector/config.go#L151-#L168 
[grpc-version]: https://github.com/kubernetes-sigs/apiserver-network-proxy/blob/4e8b12d2cd77db7ae5acecb5b706d5ecb3bf93e3/go.mod#L14 
[konn-loko]: https://github.com/ipochi/konnectivity-on-lokomotive

**Comment by jkh52** [comment]
I've tried to repro this with proxy-server and -agent 0.0.24, and grpc + uds modes.

My scripts are much simpler:

```
$ cat foo.sh 
#!/bin/bash

for i in {1..500}
do
   kubectl logs kube-dns-7fd8755dc6-cs8vz sidecar -n kube-system > /dev/null 2> /dev/null
done

$ cat bar.sh 
#!/bin/bash

for i in {1..50}
do
    /run/foo.sh &
done

$
```

I am unable to repro the leak so far. I intentionally chose a pod with relatively small logs payload, I'll try a larger one to see if that helps repro.

**Comment by rata** [comment]
@jkh52 thanks for trying that out. We will continue to investigate and share here any findings we find. Let us know if you have any suggestions or if you do investigate more yourself on this, it will be great! :)

**Comment by ipochi** [comment]
@jkh52 

Hi Joseph, thanks for trying it out. 

I added a bit of randomness to the scripts you shared and increased the concurrency. I had only nginx deployment in the default namespace (25 replicas)

```
./bar.sh
```


```
cat foo.sh

#!/bin/bash

NAMESPACE=${NAMESPACE:-"default"}
pods=$(kubectl get pods -n ${NAMESPACE} --no-headers | awk '{print $1}')
FLOOR=1;
arr=(`echo ${pods}`);
CEILING=${#arr[@]};

for i in {1..2}
do
   RANGE=$(($CEILING-$FLOOR));
   random=$RANDOM;
   let "random %= $RANGE";
   random=$(($random+$FLOOR));
   kubectl logs -n ${NAMESPACE} ${arr[${random}]} > /dev/null 2> /dev/null
done
```

and 

```
cat bar.sh
#!/bin/bash

set -euo pipefail

for i in {1..500}
do
    ./foo.sh &
done
```

I can see FD's leaking, for example post the end of the script, the number of open file descriptors were `1528`

**Comment by ipochi** [comment]
Hello, 

Upon further investigation on the issue we discovered that the possible reason for file descriptor/memory leak could be the idle connections not freeing up the resources.

We tried this on a local setup with debug statements added to collect more information about the connections i.e PacketType’s being passed for each connection id ( e.g PacketType_DIAL_REQ, PacketType_DIAL_RES, PacketType_CLOSE_REQ, PacketType_CLOSE_RSP).

In one of the scenarios from a total of 1470 connection id’s, there were about 133 connections  that are probably leaking whose ids never send a `PacketType_CLOSE_REQ` from the kube-apiserver and consequently never receive `PacketType_CLOSE_RSP` packet. This leads us to further confirm the understanding that idle connections are at play here that are not freeing up the resources. Also, it seems there is no attempt to close the connections, it is not that some packet is being lost, but we only try to close the connections as part of the defer call which will never happen if the goroutine is stuck.


To further confirm that indeed is the case, the value 131 of `konnectivity_network_proxy_server_grpc_connections{service_method="Proxy"}` metric seemed to be in sync with the number of connection ids in idle state.

Similar synchronization between the two metrics was noted in different runs

I am attaching the goroutine profile from a previous run and today’s run, in both runs the proxy-server was bombarded with random traffic of logs and port-forward connections. The percentage (~16% in the Proxy() method) of goroutines in the first block of Proxy method is strikingly similar in both runs.

![image](https://user-images.githubusercontent.com/6972950/141153629-405c2283-7f59-4dfd-8472-0f67b5a23d77.png)

![image](https://user-images.githubusercontent.com/6972950/141153819-7f4aae6d-1090-4c6f-998a-10e914cda40e.png)




Also attached is the heap profile for one of the test runs, which show about 90% of memory is used in the gRPC functions possibly as a result of idle gRPC connections. 

![image](https://user-images.githubusercontent.com/6972950/141154096-28f06b4a-b98a-4bfe-88aa-8e9403b5b982.png)


/cc @iaguis @rata

**Comment by ipochi** [comment]
Traffic was generated with the script using kubectl [above](https://github.com/kubernetes-sigs/apiserver-network-proxy/issues/276#issuecomment-957641007) and a [cli tool](https://github.com/ipochi/konnscen), getting same results in both.

**Comment by rata** [comment]
Here is a summary of the current history, based on what @ipochi just mentioned.

To reproduce this issue using kubectl in a cluster with konnectivity using GRPC and UDS, you can do the following:
```
kubectl -n kube-system run nginx --image=nginx --replicas=30
$ cat run.sh 
#!/bin/bash

set -euo pipefail

for i in {1..500}
do
    ./get-logs.sh &
done

$ cat get-logs.sh 
#!/bin/bash

#NAMESPACE=${NAMESPACE:-"default"}
NAMESPACE=${NAMESPACE:-"kube-system"}
pods=$(kubectl get pods -n ${NAMESPACE} --no-headers | awk '{print $1}')
FLOOR=1;
arr=(`echo ${pods}`);
CEILING=${#arr[@]};

for i in {1..2}
do
   RANGE=$(($CEILING-$FLOOR));
   random=$RANDOM;
   let "random %= $RANGE";
   random=$(($random+$FLOOR));
   kubectl logs -n ${NAMESPACE} ${arr[${random}]} > /dev/null 2> /dev/null
done

# Once you have both of this files with exec permissions
./run.sh
```

After this finishes running you will see the prom metric in proxy-server (go to `http://<proxy-server-ip>:<admin-port>/metrics`):
```
konnectivity_network_proxy_server_grpc_connections{service_method="Proxy"} XX
process_open_fds XX
```

The mem usage of proxy-server will increase as you run this more times.


 I've been looking at what can be the problem and I have not really found it. There are several things I tried that I'm not sure why it happens:
1. If I try adding [here][uds-keepalive] the [MaxConnectionIdle][keepalive-doc], it doesn't close idle connections that are waiting in stream.Recv(). I tried adding that param manually to 2 minutes, but grpc connections seem open after that looking at the prom metrics and open fds. Do you happen to know what can cause MaxConnectionIdle to not close the seemingly idle connections? I'm not very familiar with the proto messages that the Proxy method sends/recvs in a tream, so maybe it is something related to that?
1. The grpc connection created from the APIserver to proxy-server is closed by the apiserver when [this go routine finishes running][grpc-close] (not sure when the `Close()` method is invoked yet). I tried adding logs on errors paths, this is how I come up with PR #290. However, I find it weird that when I only run kubectl logs once, the command finishes and shows the logs correctly, but the connection between the APIserver and proxy-server is killed after 1 minute and 30 seconds. I guess that the keepalive is kicking in (I'm using 1 min for the [time param here][uds-keepalive]). Is this expected? Shouldn't [the konn-client Close() method][konn-close] be called? Do you have any clues on where this is used from the API server?
1. Even though the MaxConnIdle param from 1 didn't work, I ended up trying [this ugly hack][kube-hack] to make sure these are all the connections established and I can close them. I did it in the kube repo in the vendor subdir, to easily test it. With this horrible patch, I see the prom metric `konnectivity_network_proxy_server_grpc_connections{service_method="Proxy"}` go back to 0, so that is good. Although I see several hundred open fds, so maybe this hack is not working even as a hack that well. Will try to debug this next week, it is quite late here now :)

If you have any hints, please let me know. It is quite hard to make progress on this, some eyes more familiar with the code can definitely help!

Let me know if I can help you to reproduce this or anything else too :) 

[uds-keepalive]: https://github.com/kubernetes-sigs/apiserver-network-proxy/blob/fc56f4ba47a3da62519f93743830f13abf962492/cmd/server/app/server.go#L189
[keepalive-doc]: https://pkg.go.dev/google.golang.org/grpc/keepalive#ServerParameters
[grpc-close]: https://github.com/kubernetes-sigs/apiserver-network-proxy/blob/fc56f4ba47a3da62519f93743830f13abf962492/konnectivity-client/pkg/client/client.go#L95
[konn-close]: https://github.com/kubernetes-sigs/apiserver-network-proxy/blob/fc56f4ba47a3da62519f93743830f13abf962492/konnectivity-client/pkg/client/conn.go#L115
[kube-hack]: https://github.com/kinvolk/kubernetes/commit/2a190c6a42b5fbb70d368a141422d7f667385808

**Comment by rata** [comment]
Oh, forgot to add that I also tested this with latest k8s (that uses grpc 1.41) and it doesn't help. I also tried with 1.42 and same results.

However, opened a PR to upgrade grpc here to 1.42, just to stay with the latest fixes: https://github.com/kubernetes-sigs/apiserver-network-proxy/pull/291

**Comment by rata** [comment]
With @290 in, I think the remaining steps to close this issue are:
 * Make the improvements to #290  mentioned [here][link]
 * Update k/k to use either 0.0.26 that includes #290 or 0.0.27 with the fixes in ^ too (once we have them :))
 * Catch the other leak. With this fix, instead of leaking 60 or sometimes 600 fds (depends on the run), it seems we know leak ~10 fds (haven't done enough runs yet to determine if this is always the case, but so far it seems it is). We need to hunt that down and fix it. I have the gut feeling that this might be in common code with http-connect, so I'd like to try the repro for this issue with http-connect and see if a similar amount of fds is leaked. Maybe some call to [close][close] is missing in some part or something like that. But we need to investigate this.
 * I'd like to see if we can add some "big hammer" to mitigate this kind of leaks in the future, so we stop the bleeding while we fix the underlying issue. I'm thinking something along the lines of adding an optional flag to proxy-server to close connections that are inactive after X duration (e.g. `--grpc-close-inactive-conn Xh`). Grpc keepalive params it seems can't achieve this, but will look for other grpc params just in case (if not we can easily do it ourselves too. It should be similar to the patches I linked already in the issue). Right now connections can be kept open for days, and they consume lot of memory. If we add a way to close inactive connections, we should be able to use konnectivity with kube versions that don't include #290 or mitigate the impact of a leak if we regress in the future. In our case, we had to stop the rollout of this due to the leak we found. It will be great if in the future we can add some flag if needed to stop the bleeding while we work on a fix for the issue, instead of stopping the rollout.

Let me know what you think, specially about the last item :)

[link]: https://github.com/kubernetes-sigs/apiserver-network-proxy/pull/290#discussion_r750557397
[close]: https://github.com/kubernetes-sigs/apiserver-network-proxy/blob/5944ae0ba9ebf5ae14a3cc412a337d29794ccb50/konnectivity-client/pkg/client/conn.go#L115

**Comment by jkh52** [comment]
> Update k/k to use 0.0.26

@jdnurme started this at https://github.com/kubernetes/kubernetes/pull/106634. That PR could use some help from anyone that understands hack/vendor-update.sh issues.

**Comment by ipochi** [comment]
Solution to terminate idle connections and stop leaking of file descriptors is proposed via #312.

Instead of terminating idle connections by settings `MaxConnectionAge` and `MaxconnectionAgeGrace` which would close all grpc connections regardless of the activity status, this solution only closes those connections/stream that are lying idle consuming resources.

**Comment by andrewsykim** [comment]
@ipochi @rata I've been trying to reproduce this issue but no luck, maybe you can shed some light and help me reproduce it.

Here's the script I'm using, it's mostly identical to the original scripts you shared with some minor tweaks:
```bash
$ cat get-logs.sh 
#!/bin/bash

NAMESPACE=${NAMESPACE:-"kube-system"}
pods=$(kubectl get pods -n ${NAMESPACE} --no-headers | awk '{print $1}')
FLOOR=1;
arr=(`echo ${pods}`);
CEILING=${#arr[@]};

echo $pods

for i in {1..10}
do
   RANGE=$(($CEILING-$FLOOR));
   random=$RANDOM;
   let "random %= $RANGE";
   random=$(($random+$FLOOR));
   kubectl logs -n ${NAMESPACE} ${arr[${random}]} > /dev/null 2> /dev/null
done

$ cat konnectivity-debug.sh 
#!/bin/bash

set -euo pipefail

for i in {1..100}
do
    ./get-logs.sh &
done
```

After running this a few times, I checked the metrics endpoint of proxy server :
```
$ curl <proxy endpoint>/metrics | grep 'konnectivity_network_proxy_server_grpc_connections{service_method="Proxy"}'
konnectivity_network_proxy_server_grpc_connections{service_method="Proxy"} 4

$ curl <proxy endpoint>:8134/metrics | grep 'process_open_fds'

# HELP process_open_fds Number of open file descriptors.
# TYPE process_open_fds gauge
process_open_fds 22
```

Checking /proc/<pid>/fd directly also shows similar results.

**Comment by rata** [comment]
@andrewsykim sorry for the delay, I was sick :(

I think you are seeing the issue. Probably you are running from latest master or a tree that includes #290 and since that commit the fd leak is not _so_ big. But you are seeing 4 `konnectivity_network_proxy_server_grpc_connections` and 22 fds, that is the leak. The `konnectivity_network_proxy_server_grpc_connections` should be 0 (at least after some time). If you run that again, I bet you will see that metric go to 7-8. It will only increase each time you run that script that just gets logs, even if you leave it running for days.

The same is observed in the open_fds metric. You are seeing 22 fds, usually there are about 13 at startup. Doing an `ls -l` in /proc/pid/fd you can see the fds belong to the connectivity socket (see the original report on what I did to go from the fd inode to know which connection it is being used for).

Let me know if you think otherwise, but I think you reproduced it. Just that the remaining leak, after the #290 fix, is not that big.

Also, the script you are using seems correct and almost identical to mine.

btw, @ipochi told me that using http-connect mode the increase in fds and the corresponding prometheus metric is not seen, all goes back to the original numbers as expected.

**Comment by andrewsykim** [comment]
An interesting observation I noticed while trying to reproduce this issue is that restarting Konnectivity Server will not completely clear the open files and goroutines. So if I restart just proxy server and immediately check the goroutine and open files count, it's a bit lower but still very high. But if I restart kube-apiserver, I see the goroutines and open files reset across both components. So at this point I'm thinking this is mainly a leak in konnectivity-client. 

Going to run some validation against https://github.com/kubernetes-sigs/apiserver-network-proxy/pull/325 today which has some potential fixes in konnectivity-client but would appreciate it if anyone else that is able to reproduce the issue can also test it. Note that the fix does require a rebuild of kube-apiserver so it might be difficult to test it

**Comment by rata** [comment]
@andrewsykim thanks! We've noticed the same (if you dig into the comments, I think me or @ipochi mentioned it). I think we've seen other weird things too, like if you kill the proxy-server for some minutes, then you start it again and it start with even more fds than just before killing it (it seems like konnectivity client is retrying with more connections, until the proxy server is up and all the connections gets established or something like that).

I'll try it next week with the kube-apiserver modifications. Hopefully one of those fixes also fixes this! :crossed_fingers:

**Comment by andrewsykim** [comment]
> @andrewsykim thanks! We've noticed the same (if you dig into the comments, I think me or @ipochi mentioned it). I think we've seen other weird things too, like if you kill the proxy-server for some minutes, then you start it again and it start with even more fds than just before killing it (it seems like konnectivity client is retrying with more connections, until the proxy server is up and all the connections gets established or something like that).

Yeah I see this too, and sorry for missing your comments above.

**Comment by andrewsykim** [comment]
Ran some more tests against https://github.com/kubernetes-sigs/apiserver-network-proxy/pull/325 which wasn't very promising.  

However, I got some pretty promising results from https://github.com/kubernetes/kubernetes/pull/108215, which does seem to explain a lot of behaviors we're seeing when proxy-server is restarted. Reposting from the PR:

After some investigation and testing, the goroutine leak seems to be happening because the gRPC dialer in kube-apiserver is non-blocking by default. So when proxy server becomes unavailable, new connections requiring the dialer (pod/exec, pod/logs, webhook requests, etc) would spin up a goroutine in the background and try to connect. And because we passed context.TODO() into CreateSingleUseGrpcTunnel(), the tunnel continued to run even though the request was likely cancelled already.

**Comment by rata** [comment]
@andrewsykim thanks! That might help, I was trying it with a local kube cluster here but with the dockershim deprecation and some CNI issue, it is not starting. I'll have a closer look, but if not I might try to run an actual cluster with those images.

**Comment by rata** [comment]
@andrewsykim so, you see the `konnectivity_network_proxy_server_grpc_connections{service_method="Proxy"}` and `process_open_fds` stay the same after/before running the repro [here][link]?

[link]: https://github.com/kubernetes-sigs/apiserver-network-proxy/issues/276#issuecomment-1022797259

**Comment by rata** [comment]
Hmm, that seems like a leak. The `konnectivity_network_proxy_server_grpc_connections{service_method="Proxy"}` metric should return to 0 if we are not tunneling any connection. We need to wait approx 1.5 min, so some timeout kicks in and kills some, but that seems to be leaking grpc connections :(

The number of open fds is also quite high (at startup is aroun 13). I bet that if you look at the fds in /proc/pid/fd, see the inode and see what that is used for, that is the konnectivity UDS (see the original bug report to see how to go from fd to inode and which connection is using it. It is quite simple :)).

Can you confirm if you see that? IIUC and you see that, it is still leaking :(

**Comment by andrewsykim** [comment]
> Hmm, that seems like a leak. The konnectivity_network_proxy_server_grpc_connections{service_method="Proxy"} metric should return to 0 if we are not tunneling any connection. We need to wait approx 1.5 min, so some timeout kicks in and kills some, but that seems to be leaking grpc connections :(

Sorry, when I shared those metrics I was still running tests on the cluster. These are the metrics after the tests have stopped for a bit:
```
$ curl localhost:<admin-port>/metrics | grep konnectivity_network_proxy_server_grpc_connections

# HELP konnectivity_network_proxy_server_grpc_connections Number of current grpc connections, partitioned by service method.
# TYPE konnectivity_network_proxy_server_grpc_connections gauge
konnectivity_network_proxy_server_grpc_connections{service_method="Connect"} 5
konnectivity_network_proxy_server_grpc_connections{service_method="Proxy"} 2

$ curl localhost:8134/metrics | grep open

# HELP process_max_fds Maximum number of open file descriptors.
# HELP process_open_fds Number of open file descriptors.
# TYPE process_open_fds gauge
process_open_fds 20
```

Note that there are some requests that are still going through the proxy that I didn't disable, like metrics server requests

**Comment by rata** [comment]
@andrewsykim No, none seem to fix this issue at all :(

I tested with patches:
 * https://github.com/kubernetes/kubernetes/pull/108215
 * #328 
 * #325 
 
I used ANP master and applied the aforementioned konnectivity-client patches to the k8s build. The build is latest k8s master, so it does include #290 too. 

If I run the script [here][link], I see the `konnectivity_network_proxy_server_grpc_connections{service_method="Proxy"}` increase each time I run it (it is about 3 sometimes, then I run it again after waiting 3 minutes and it goes up to 6, sometimes it increases in about 10 in after a run finished, etc.). I can make that number arbitrary high, is just a matter of running the run.sh repro script enough times.

The same happens with the process_open_fds metric, it always increases.

@andrewsykim I guess you can repro this same behavior in your setup, right? (always increasing after each run of the run.sh script and waiting 3 minutes to run it again, so some timout kicks in after 1:30m)

If I shutdown the proxy now, though, it all goes back to 0. So, in that regards, it seems that https://github.com/kubernetes/kubernetes/pull/108215 is doing what we expect. But the leak we reported is still here, with all these patches applied.

[link]: https://github.com/kubernetes-sigs/apiserver-network-proxy/issues/276#issuecomment-1022797259



The diff on my k8s repo is this, just in case and using k8s commit d5c26190ac84a3795526bfe2c11df43a00321301 as base:
```
diff --git a/staging/src/k8s.io/apiserver/pkg/server/egressselector/egress_selector.go b/staging/src/k8s.io/apiserver/pkg/server/egressselector/egress_selector.go
index 9da0e2a099c..b93a0ce9825 100644
--- a/staging/src/k8s.io/apiserver/pkg/server/egressselector/egress_selector.go
+++ b/staging/src/k8s.io/apiserver/pkg/server/egressselector/egress_selector.go
@@ -200,7 +200,7 @@ func (u *udsGRPCConnector) connect() (proxier, error) {
        })
 
        ctx := context.TODO()
-       tunnel, err := client.CreateSingleUseGrpcTunnel(ctx, udsName, dialOption, grpc.WithInsecure())
+       tunnel, err := client.CreateSingleUseGrpcTunnel(ctx, udsName, dialOption, grpc.WithInsecure(), grpc.WithBlock())
        if err != nil {
                return nil, err
        }
diff --git a/vendor/sigs.k8s.io/apiserver-network-proxy/konnectivity-client/pkg/client/client.go b/vendor/sigs.k8s.io/apiserver-network-proxy/konnectivity-client/pkg/client/client.go
index 6f247128def..aa79dc73f40 100644
--- a/vendor/sigs.k8s.io/apiserver-network-proxy/konnectivity-client/pkg/client/client.go
+++ b/vendor/sigs.k8s.io/apiserver-network-proxy/konnectivity-client/pkg/client/client.go
@@ -76,6 +76,7 @@ func CreateSingleUseGrpcTunnel(ctx context.Context, address string, opts ...grpc
 
        stream, err := grpcClient.Proxy(ctx)
        if err != nil {
+               c.Close()
                return nil, err
        }
 
@@ -161,6 +162,7 @@ func (t *grpcTunnel) serve(c clientConn) {
                        t.connsLock.RUnlock()
 
                        if ok {
+                               conn.closed = true
                                close(conn.readCh)
                                conn.closeCh <- resp.Error
                                close(conn.closeCh)
diff --git a/vendor/sigs.k8s.io/apiserver-network-proxy/konnectivity-client/pkg/client/conn.go b/vendor/sigs.k8s.io/apiserver-network-proxy/konnectivity-client/pkg/client/conn.go
index cc6e66be237..85bf8f7b515 100644
--- a/vendor/sigs.k8s.io/apiserver-network-proxy/konnectivity-client/pkg/client/conn.go
+++ b/vendor/sigs.k8s.io/apiserver-network-proxy/konnectivity-client/pkg/client/conn.go
@@ -38,6 +38,7 @@ type conn struct {
        random  int64
        readCh  chan []byte
        closeCh chan string
+       closed  bool
        rdata   []byte
 }
 
@@ -114,6 +115,16 @@ func (c *conn) SetWriteDeadline(t time.Time) error {
 // proxy service to notify remote to drop the connection.
 func (c *conn) Close() error {
        klog.V(4).Infoln("closing connection")
+       defer func() {
+               // We need to close any channels in the case where
+               // the CLOSE_REQ was not sent or the CLOSE_RSP was
+               // not received. This is usually the case when the
+               // transport is already closed.
+               if !c.closed {
+                       close(c.readCh)
+               }
+       }()
+
        var req *client.Packet
        if c.connID != 0 {
                req = &client.Packet{

```

**Comment by rata** [comment]
Also, I wonder that if [this comment from @cheftako ][link] that is still not addressed, in particular the item 1 that is mentioned, might be the remaining leak (not real evidence, just a hunch). After all, that fix in #290 reduced this leak _a lot_, if there are associated connections that needs to be closed in that case too, it might be the remaining leak.

@andrewsykim I'm not sure how the patch fixing that should look like, but maybe it is simpler for you to figure it out?


[link]: https://github.com/kubernetes-sigs/apiserver-network-proxy/pull/290#discussion_r749947809

**Comment by andrewsykim** [comment]
Thanks for reporting back @rata. Yesterday I was also able to reproduce some remaining leaks with those patches as well but not clear yet if it's the same leak you're seeing. I'll look into Walter's comment in that PR as well. 

The leak I reproduced yesterday was when I was trying to reproduce the issue described here https://github.com/kubernetes-sigs/apiserver-network-proxy/pull/324. From the pprof dump, the goroutine leak was mainly happening due to semaphore lock when sending to the agent stream:
```
goroutine 192765 [semacquire, 125 minutes]:
sync.runtime_SemacquireMutex(0xc00038c094, 0x1428700, 0x1)
        /usr/local/go/src/runtime/sema.go:71 +0x47
sync.(*Mutex).lockSlow(0xc00038c090)
        /usr/local/go/src/sync/mutex.go:138 +0x105
sync.(*Mutex).Lock(...)
        /usr/local/go/src/sync/mutex.go:81
sigs.k8s.io/apiserver-network-proxy/pkg/server.(*backend).Send(0xc00038c090, 0xc003b6fb00, 0x0, 0x0)
        /go/src/sigs.k8s.io/apiserver-network-proxy/pkg/server/backend_manager.go:86 +0xb9
sigs.k8s.io/apiserver-network-proxy/pkg/server.(*ProxyServer).serveRecvFrontend(0xc000283b00, 0x190d458, 0xc003a5d930, 0xc0072ddc20)
        /go/src/sigs.k8s.io/apiserver-network-proxy/pkg/server/server.go:432 +0xacc
created by sigs.k8s.io/apiserver-network-proxy/pkg/server.(*ProxyServer).Proxy
        /go/src/sigs.k8s.io/apiserver-network-proxy/pkg/server/server.go:367 +0x318
```

@rata if you're environment is still available, can you get the goroutine stack dump from proxy server so we can check if it's the same leak. Here are the steps to get it:
```
curl localhost:<admin-port>/debug/pprof/goroutine?debug=2
```

**Comment by rata** [comment]
@andrewsykim thank you! Keep us posted if you have any other findings, please!

I also have 1m as keepalive, I'm starting the proxy-server like this:

```
./bin/proxy-server --server-port=0 --uds-name=/tmp/konnectivity-server.socket --cluster-ca-cert=certs/agent/issued/ca.crt --cluster-cert=certs/agent/issued/proxy-frontend.crt --cluster-key=certs/agent/private/proxy-frontend.key --warn-on-channel-limit --v 2 --keepalive-time 1m --frontend-keepalive-time 1m  --enable-profiling
```

I don't see that semaphore stucked here. Here is the full dump using that curl command, in case I'm missing something: https://pastebin.com/ptZ3jDqB


I've run pprof for the goroutines and found more evidence that the go-routines are just sitting there waiting to receive a pkg (like I said in the original report). When the `konnectivity_network_proxy_server_grpc_connections{service_method="Proxy"}` is 6, I see 6 goroutines in ProxyServer.ServeRecvFrontend. When I run the repro again, the metric goes up (in this case to 12) and I see 12 go routines in that function. So, it really seems like either the other apiserver/konn-client is not closing the connection with proxy server, therefore proxy-server [received and EOF](https://github.com/kubernetes-sigs/apiserver-network-proxy/blob/d2eec1ec3ac33f683c81a950f171caab568c0250/pkg/server/server.go#L384) and therefore never return (on return it [closes the recvCh](https://github.com/kubernetes-sigs/apiserver-network-proxy/blob/d2eec1ec3ac33f683c81a950f171caab568c0250/pkg/server/server.go#L375-L378) so the goroutine in serveRecvFrontend should stop) or that stream.Recv() call is getting stucked in a broken connection or something like that.

I think the getting stucked is not unlikely, specially as this can't take a context. But the things we tried about that didn't work. But connection not being closed by konn-client/apiserver is not unlikely either, as #290 fixed one case of that.

Also, if I grep the apiserver logs for "client.go" (the konnectivity-client file is named like that), I only see log lines like this:
```
I0224 17:13:44.273971 1066142 client.go:118] "DialResp not recognized; dropped" connectionID=1661 dialID=908622399869719253
```

All the other cases in the konn-client that don't have a return, have a log that should be printed. So, it seems we are not hitting any of those cases either.

I think the other explanation for this issue might be related to the comment @cheftako did on #290. Maybe there is something like that going on that explains the remaining leak?

I attach the pprof image for the leaking goroutines. First when they were 6 leaking (as reported by `konnectivity_network_proxy_server_grpc_connections{service_method="Proxy"}`)

![gorotine-6-leaks](https://user-images.githubusercontent.com/70861/155566775-17bd81b1-b9a5-4945-b1d8-1b14b4b011ee.png)

And this when the leaking goroutine were 12:

![gorotine-12-leaks](https://user-images.githubusercontent.com/70861/155566782-008d1d9f-9000-4cb1-ad1a-1dc867aaa257.png)

**Comment by andrewsykim** [comment]
> So, it really seems like either the other apiserver/konn-client (it might be the apiserver code not closing a connection in some specific case, outside of konn-client) is not closing the connection with proxy server, therefore proxy-server [received and EOF](https://github.com/kubernetes-sigs/apiserver-network-proxy/blob/d2eec1ec3ac33f683c81a950f171caab568c0250/pkg/server/server.go#L384) and therefore never return (on return it [closes the recvCh](https://github.com/kubernetes-sigs/apiserver-network-proxy/blob/d2eec1ec3ac33f683c81a950f171caab568c0250/pkg/server/server.go#L375-L378) so the goroutine in serveRecvFrontend should stop) or that stream.Recv() call is getting stucked in a broken connection or something like that.

The former might be expected if kube-apiserver doesn't call Close() on the connection, which is dependent on various things like default keepalives for a connection and how client-go handles net.Conn from `Dial()`. Maybe one way to rule this out is to restart kube-apiserver and see if that resets all the connections for proxy server too. If the issue was stream.Recv() getting stuck, it probably wouldn't matter if the other side of the stream restarted. Are you able to test this?

**Comment by rata** [comment]
@andrewsykim good idea. Killing the apiserver makes the metric go down back to 0: `konnectivity_network_proxy_server_grpc_connections{service_method="Proxy"} 0`. So, probably a missing close on the konn-client or apiserver side is what we are after here.

Just curious, did you manage to repro this, then?

One easy way you can try, is the following (@ipochi realized about just using the proxy-server in the host. I had issues trying to use the cluster certs in the past):
- Clone k8s repo and start a cluster with `hack/local-up-cluster.sh`. You can use this env var to have an egress selector in the started cluster: export EGRESS_SELECTOR_CONFIG_FILE=path.yaml
- Just start proxy-server as a processes in the host with: `./bin/proxy-server --server-port=0 --uds-name=/tmp/konnectivity-server.socket --cluster-ca-cert=certs/agent/issued/ca.crt --cluster-cert=certs/agent/issued/proxy-frontend.crt --cluster-key=certs/agent/private/proxy-frontend.key --warn-on-channel-limit --v 2 --keepalive-time 1m --frontend-keepalive-time 1m  --enable-profiling`
- And start the agent as a process in the host too: ./bin/proxy-agent --ca-cert=certs/agent/issued/ca.crt --agent-cert=certs/agent/issued/proxy-agent.crt --agent-key=certs/agent/private/proxy-agent.key
- Make sure your egress selector uses the proper UDS path if you adjust.

Then, as the k8s doc says, you can `export KUBECONFIG=/var/run/kubernetes/admin.kubeconfig` and just use kubectl, so it should be good to use with the run.sh script. 

In this setup, you should easily see it. And if you want to try some changes to the apiserver, you can just modify the code (or the vendored code if you want to change the konn-client), run `hack/local-up-cluster.sh` to compile k8s and start the cluster.

Let me know if you can't repro or something :)

**Comment by andrewsykim** [comment]
> Just curious, did you manage to repro this, then?

I'm not 100% sure to be honest -- I have seen a handful of connections lingering around but I can't determine if it's expected due to some on-going webhook requests (like metrics server) or due to a leak. This is also going into some client-go level details that are hard to debug. I will try the steps you mentioned to try to reproduce the issue.

**Comment by andrewsykim** [comment]
> Clone k8s repo and start a cluster with hack/local-up-cluster.sh. You can use this env var to have an egress selector in the started cluster: export EGRESS_SELECTOR_CONFIG_FILE=path.yaml

Can you share the egress selector config you used here?

**Comment by andrewsykim** [comment]
@rata I'm able to reproduce your leak now, thanks for the detailed steps.

If I run run.sh a few times, I do see goroutine/connections created that are never cleaned up. I think I was also able to pin point where the leak is as well, but would appreciate your help validating the fix.

It seems like there is a race condition that could happen when receiving from the pending dial channel after sending a dial request and when a context cancel is received from DialContext. And the race condition is specifically possible because we use unbuffered channels for the pending dial channel. The sequence goes something like this:

1. When grpcTunnel.DialContext() is called, we add an unbuffered channel to pendingDials and send the first DIAL_REQ to proxy server from kube-apiserver:

https://github.com/kubernetes-sigs/apiserver-network-proxy/blob/d2eec1ec3ac33f683c81a950f171caab568c0250/konnectivity-client/pkg/client/client.go#L186-L189

2. In parallel, kube-apiserver is also receiving on the grpc stream, looking for the first DIAL_RSP:

https://github.com/kubernetes-sigs/apiserver-network-proxy/blob/d2eec1ec3ac33f683c81a950f171caab568c0250/konnectivity-client/pkg/client/client.go#L112-L131

3. The channel for pendingDial is removed using a defer function in DialContext, meaning it would be removed on dial timeout, or on context cancel:

https://github.com/kubernetes-sigs/apiserver-network-proxy/blob/d2eec1ec3ac33f683c81a950f171caab568c0250/konnectivity-client/pkg/client/client.go#L190-L194

https://github.com/kubernetes-sigs/apiserver-network-proxy/blob/d2eec1ec3ac33f683c81a950f171caab568c0250/konnectivity-client/pkg/client/client.go#L228-L232

4. It is possible that kube-apiserver receives a DIAL_RSP (based on DIAL_REQ sent from DialContext), before the context passed into DialContext is about to be cancelled. In which case, it recognizes the pendingDial and sends the dial result to the channel. And because the channel is unbuffered, it receives the message even though DialContext is no logner reading from the pending dial channel: 

https://github.com/kubernetes-sigs/apiserver-network-proxy/blob/d2eec1ec3ac33f683c81a950f171caab568c0250/konnectivity-client/pkg/client/client.go#L125-L131

5. Afterwards, the goroutine continues to block on stream.Recv() at the top of grpcTunnel.serve() thinking the dial result was sent to the pending dial channel. But because nothing is reading the unbuffered channel, the connection would never be established here:

https://github.com/kubernetes-sigs/apiserver-network-proxy/blob/d2eec1ec3ac33f683c81a950f171caab568c0250/konnectivity-client/pkg/client/client.go#L218-L227

**Comment by andrewsykim** [comment]
From my initial testing so far, it seems like just using an unbuffered channel that blocks on receive works: https://github.com/kubernetes-sigs/apiserver-network-proxy/pull/337. 

I tested this by running ./run.sh multiple times and seeing the Proxy connections always goes down to 0 after a couple of minutes:
```
$ curl localhost:8095/metrics | grep grpc
# TYPE konnectivity_network_proxy_server_grpc_connections gauge
konnectivity_network_proxy_server_grpc_connections{service_method="Connect"} 1
konnectivity_network_proxy_server_grpc_connections{service_method="Proxy"} 0
```

Interesting note: regardless of # of requests sent to kube-apiserver, there are always ~2 proxy connections that stick around longer than others. And we know this is not a leak because you can see kube-apiserver sending the CLOSE_REQ consistently after ~2 minutes. It's probably some connection pooling logic in kube-apiserver but thought it was worth noting.

**Comment by rata** [comment]
@andrewsykim The explanation makes perfect sense to me, and indeed the PR seems to fix this bug for me too! Thanks a lot, great catch!

I also always saw the 2 connection that stay longer (I think I mentioned in this issue before too), but I never saw those become a problem, so haven't stopped to debug why that behavior happens. If you do, please let me know, I'm curious. It seems to be quite close to a grpc keepalive timeout (like 1m and 30s, although 1m and 20 seconds will really match the defaults), but I haven't checked it  :)

One weird thing that I saw, though, is that once the apiserver crashed. I added #337 to the k8s vendored code with the changes I already have here (#328 and #325). The output of the kube-apiserver logs is this when it crashed:

```
panic: close of closed channel

goroutine 34197 [running]:
k8s.io/kubernetes/vendor/sigs.k8s.io/apiserver-network-proxy/konnectivity-client/pkg/client.(*grpcTunnel).serve(0xc01046ade0, {0x53fd8e0, 0xc0103b0a80})
        /home/rodrigo/src/kinvolk/kubernetes/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/sigs.k8s.io/apiserver-network-proxy/konnectivity-client/pkg/client/client.go:166 +0x9a5
created by k8s.io/kubernetes/vendor/sigs.k8s.io/apiserver-network-proxy/konnectivity-client/pkg/client.CreateSingleUseGrpcTunnel
        /home/rodrigo/src/kinvolk/kubernetes/kubernetes/_output/local/go/src/k8s.io/kubernetes/vendor/sigs.k8s.io/apiserver-network-proxy/konnectivity-client/pkg/client/client.go:90 +0x196

```

However, after that I've recompiled kubernetes with no changes (I mean, running the hack/local-up-cluster script), run this again several times and couldn't hit it again. I did this complete procedure several times.

In my vendored konnectivity-client/pkg/client/client.go line 166 points to this close: https://github.com/kubernetes-sigs/apiserver-network-proxy/pull/325/files#diff-a37bb159e5baec4208de32e78a3905ecef6282ceaf095115842e4e3738d96310L164 (it is line 165 with that PR diff currently, but if you apply #328 too then it adds one more line and becomes line 166).

So, PR #337 seems good to me, fixes the issue and doesn't seem related to the double close crash. However, I'm skeptical about #325 that might introduce a case where a double close is possible, but just very hard to hit. But I'm not sure, as I said, I could not reproduced it ever again.

**Comment by andrewsykim** [comment]
Thanks for testing @rata, and again thanks for the reproduction steps. 

> So, PR https://github.com/kubernetes-sigs/apiserver-network-proxy/pull/337 seems good to me, fixes the issue and doesn't seem related to the double close crash. However, I'm skeptical about https://github.com/kubernetes-sigs/apiserver-network-proxy/pull/325 that might introduce a case where a double close is possible, but just very hard to hit. But I'm not sure, as I said, I could not reproduced it ever again.

After thinking about this PR more, I don't think we need it because the goroutine leak was only reproducible in the test because the http client in the test was not calling conn.Close(). Will close it for now

**Comment by rata** [comment]
I'd like to keep this open, though (or I can open a new one, the same for me) to tackle some of the other things listed [here](https://github.com/kubernetes-sigs/apiserver-network-proxy/issues/276#issuecomment-974141778).

I'm interest to know if we can add some "big hammer" to kill inactive connections and work-around these leaks from proxy-server side. This will help in cases where you need to run k8s versions that don't yet have the fix.

The other thing listed there, not yet done, is to check why these 2 open connections always take longer to finish, that we were talking about @andrewsykim :). I consider that more "nice to have" :)

**Comment by andrewsykim** [comment]
> I'd like to keep this open, though (or I can open a new one, the same for me) to tackle some of the other things listed https://github.com/kubernetes-sigs/apiserver-network-proxy/issues/276#issuecomment-974141778.

@rata let's open a new issue for any new/remaining leaks that come up since they may not be specific to the gRPC mode w/ UDS.

**Comment by andrewsykim** [comment]
@rata closing this now that https://github.com/kubernetes-sigs/apiserver-network-proxy/pull/337 is merged. I'll also work on the k/k bump as soon as the release is cut. Please feel free to open a new issue for the other issues you mentioned above. 

/close

**Comment by andrewsykim** [comment]
Update to Kubernetes is here https://github.com/kubernetes/kubernetes/pull/108437

**Comment by rata** [comment]
For completeness, this is for k8s master: https://github.com/kubernetes/kubernetes/pull/108437
